### PR TITLE
Decimal coerce

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 1]
```diff
@@ -399,7 +422,74 @@ private static Signature bindSignature(Signature signature, List<? extends Type>
...
+    {
+        for (TypeSignature typeSignature : signature.getArgumentTypes()) {
+            if (!typeSignature.getLiteralParameters().isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static class SignatureWithExtractedTypesWithLiteralParameters
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: maybe move this class outside with related methods (containsArgumentsWithLiteralParameters, extractTypeParametersForArgumentsWithLiteralParameters).

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 2]
```diff
@@ -399,7 +422,74 @@ private static Signature bindSignature(Signature signature, List<? extends Type>
...
+
+        public Map<String, TypeSignature> getOriginalTypeSignature()
+        {
+            return originalTypeSignature;
+        }
+    }
+
+    private static SignatureWithExtractedTypesWithLiteralParameters extractTypeParametersForArgumentsWithLiteralParameters(Signature signature)
+    {
+        if (!containsArgumentsWithLiteralParameters(signature)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: is special case required?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 3]
```diff
@@ -756,6 +854,16 @@ public static boolean isTypeOnlyCoercion(Type actualType, Type expectedType)
             return ((VarcharType) actualType).getLength().orElse(Integer.MAX_VALUE) < ((VarcharType) expectedType).getLength().orElse(Integer.MAX_VALUE);
         }
 
+        if (actualType instanceof DecimalType && expectedType instanceof DecimalType) {
+            DecimalType actualDecimal = (DecimalType) actualType;
+            DecimalType expectedDecimal = (DecimalType) expectedType;
+
+            if (actualDecimal.isShort() ^ expectedDecimal.isShort()) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:824]: why so?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 4]
```diff
@@ -756,6 +854,16 @@ public static boolean isTypeOnlyCoercion(Type actualType, Type expectedType)
...
+                return false;
+            }
+            return actualDecimal.getScale() == expectedDecimal.getScale() && actualDecimal.getPrecision() <= expectedDecimal.getPrecision();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:827]: this probably should check if dest (digits before .) >= src (digits before .) and dest.scale >= src.scale

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 5]
```diff
@@ -831,6 +939,14 @@ public static boolean isTypeOnlyCoercion(Type actualType, Type expectedType)
                 return Optional.of(new MapType(keyType.get(), valueType.get()));
             }
         }
+        if (firstType instanceof DecimalType && secondType instanceof DecimalType) {
+            DecimalType firstDecimal = (DecimalType) firstType;
+            DecimalType secondDecimal = (DecimalType) secondType;
+            int targetScale = Math.max(firstDecimal.getScale(), secondDecimal.getScale());
+            int targetPrecision = Math.max(firstDecimal.getPrecision() - firstDecimal.getScale(), secondDecimal.getPrecision() - secondDecimal.getScale()) + targetScale;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:909]: extract firstDecimal.getPrecision() - firstDecimal.getScale as firstDecimalDigitsBeforeDot. the same for second decimal

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 6]
```diff
@@ -328,15 +328,22 @@ private static boolean matchAndBind(Map<String, Type> boundParameters, Map<Strin
...
                 if (canCoerce(type, boundParameters.get(parameter.getBase()))) {
                     return true;
                 }
                 else if (canCoerce(boundParameters.get(parameter.getBase()), type) && typeParameters.get(parameter.getBase()).canBind(type)) {
-                    // Broaden the binding
+                    // Try to coerce current binding to new candidate
                     boundParameters.put(parameter.getBase(), type);
                     return true;
+                } else {
+                    // Try to use common super type of current binding and candidate
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]: this probably should only be considered only for the same base types and only for some of them (e.g. decimal). That is why I like the idea from _unification_ patch by haozun: _After this commit, coercibility between types is decoupled into 2 parts: compatibility between their based types and between their type parameters._.
That also means that method: _getCommonSuperType_ should probably be more specific or split (depending on other usages), e.g. getCoercionSuperType.

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 7]
```diff
@@ -376,12 +383,16 @@ else if (canCoerce(boundParameters.get(parameter.getBase()), type) && typeParame
 
         // The parameter is not a type parameter, so it must be a concrete type
         Type parameterType = typeManager.getType(parseTypeSignature(parameter.getBase()));
+
+        if (type.getTypeSignature().getBase().equals(parameterType.getTypeSignature().getBase())) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]: why? what does it mean?

File: presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java [hunk 8]
```diff
@@ -59,77 +59,54 @@ private static ParametricFunction decimalAddOperator()
...
-
-    public static long addShortShortShort(long a, long b, long aRescale, long bRescale)
+    public static long addShortShortShort(long a, long b)
     {
-        return a * aRescale + b * bRescale;
+        return a + b;
     }
 
-    public static Slice addShortShortLong(long a, long b, BigInteger aRescale, BigInteger bRescale)
+    public static Slice addShortShortLong(long a, long b)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java:76]: what about cross type methods (e.g. long a, Slice b)? can they be removed?

**Comment by sopel39** [comment]
lgtm % comments % check how getCommonSuperType is used (does decimal coercion work when creating array)?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java (hunk 1 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: This is not needed at all with current approach. I will just return Signature.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java (hunk 2 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: Just short circuit.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java (hunk 3 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:824]: We cannot type-only coerce decimals represented as Slice with those represented as long without

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java (hunk 4 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:827]: We do not care about digits before .

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: it's confusing, remove

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java (hunk 6 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]: I agree that supertype may mean different things in different context.
I do not quite get why only same base type should be considered.

I can leave it as is. Or for now make it work only for DECIMAL type. 
But it should probably be possible to add (SHORTINT and DECIMAL(10)).
Then I opt for leaving this as is.

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java (hunk 7 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]: It was here already

**Comment by losipiuk** [comment]
updated version after rebase
@sopel39  please take another look

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 9]
```diff
@@ -399,7 +414,45 @@ private static Signature bindSignature(Signature signature, List<? extends Type>
                 boundArguments.add(lastArgument);
             }
         }
-        return new Signature(signature.getName(), signature.getType(), bindParameters(signature.getReturnType(), boundParameters), boundArguments.build());
+        Signature boundSignature = new Signature(signature.getName(), signature.getType(), bindParameters(signature.getReturnType(), boundParameters), boundArguments.build());
+        return boundSignature;
+    }
+
+    private static boolean containsArgumentsWithLiteralParameters(Signature signature)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:421]: unused, remove

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 10]
```diff
@@ -326,18 +326,32 @@ private static boolean matchArguments(
 
     private static boolean matchAndBind(Map<String, Type> boundParameters, Map<String, TypeParameter> typeParameters, TypeSignature parameter, Type type, boolean allowCoercion, TypeManager typeManager)
     {
+        // TODO: the code flow here needs reworking. The 'if (allowCoercion()) {...
+        // do not make much sense if parameter we are matching to is calculated type.
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java:330]: does not make

**Comment by sopel39** [comment]
lgtm

**Pull request closed without merging**